tooling/lsp/src/requests/profile_run.rs

```diff
@@ -0,0 +1,104 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    future::{self, Future},
+};
+
+use acvm::{acir::circuit::Opcode, Language};
+use async_lsp::{ErrorCode, ResponseError};
+use nargo::artifacts::debug::DebugArtifact;
+use nargo_toml::{find_package_manifest, resolve_workspace_from_toml, PackageSelection};
+use noirc_driver::{CompileOptions, DebugFile, NOIR_ARTIFACT_VERSION_STRING};
+use noirc_errors::{debug_info::OpCodesCount, Location};
+
+use crate::{
+    types::{NargoProfileRunParams, NargoProfileRunResult},
+    LspState,
+};
+use fm::FileId;
+
+pub(crate) fn on_profile_run_request(
+    state: &mut LspState,
+    params: NargoProfileRunParams,
+) -> impl Future<Output = Result<NargoProfileRunResult, ResponseError>> {
+    future::ready(on_profile_run_request_inner(state, params))
+}
+
+fn on_profile_run_request_inner(
+    state: &LspState,
+    params: NargoProfileRunParams,
+) -> Result<NargoProfileRunResult, ResponseError> {
+    let root_path = state.root_path.as_deref().ok_or_else(|| {
+        ResponseError::new(ErrorCode::REQUEST_FAILED, "Could not find project root")
+    })?;
+
+    let toml_path = find_package_manifest(root_path, root_path).map_err(|err| {
+        // If we cannot find a manifest, we can't run the test
+        ResponseError::new(ErrorCode::REQUEST_FAILED, err)
+    })?;
+
+    let crate_name = params.package;
+    // let function_name = params.id.function_name();
+
+    let workspace = resolve_workspace_from_toml(
+        &toml_path,
+        PackageSelection::DefaultOrAll,
+        Some(NOIR_ARTIFACT_VERSION_STRING.to_string()),
+    )
+    .map_err(|err| {
+        // If we found a manifest, but the workspace is invalid, we raise an error about it
+        ResponseError::new(ErrorCode::REQUEST_FAILED, err)
+    })?;
+
+    // Since we filtered on crate name, this should be the only item in the iterator
+    match workspace.into_iter().next() {
+        Some(_package) => {
+            // let (mut _context, crate_id) = prepare_package(package, Box::new(get_non_stdlib_asset));
+
+            let (binary_packages, contract_packages): (Vec<_>, Vec<_>) = workspace
+                .into_iter()
+                .filter(|package| !package.is_library())
+                .cloned()
+                .partition(|package| package.is_binary());
+
+            let is_opcode_supported = |_opcode: &Opcode| true;
```

Comment:
can we get the correct backend instead?

Reply:
It's not super important but it is still relevant. May be add a todo explaining we should use the correct backend eventually?